src/dbnode/storage/bootstrap/types.go

```diff
@@ -416,5 +420,51 @@ type Source interface {
 	// A bootstrapper source should only return an error should it want to
 	// entirely cancel the bootstrapping of the node, i.e. non-recoverable
 	// situation like not being able to read from the filesystem.
-	Read(ctx context.Context, namespaces Namespaces) (NamespaceResults, error)
+	Read(ctx context.Context, namespaces Namespaces, state State) (NamespaceResults, error)
+}
+
+// InfoFileResultsPerShard maps shards to info files.
+type InfoFileResultsPerShard map[uint32][]fs.ReadInfoFileResult
+
+// InfoFilesByNamespace maps a namespace to info files grouped by shard.
+type InfoFilesByNamespace map[namespace.Metadata]InfoFileResultsPerShard
+
+// State provides a snapshot of info files for use throughout all stages of the bootstrap.
+type State struct {
+	fsOpts               fs.Options
+	infoFilesFinders     []InfoFilesFinder
+	infoFilesByNamespace InfoFilesByNamespace
+	iOpts                instrument.Options
+}
+
+// StateOptions represents the options for State.
+type StateOptions interface {
+	// Validate will validate the options and return an error if not valid.
+	Validate() error
+
+	// SetFilesystemOptions sets the filesystem options.
+	SetFilesystemOptions(value fs.Options) StateOptions
+
+	// FilesystemOptions returns the filesystem options.
+	FilesystemOptions() fs.Options
+
+	// SetInfoFilesFinders sets the finders used to load info files for all namespaces provided.
+	SetInfoFilesFinders(value []InfoFilesFinder) StateOptions
+
+	// InfoFilesFinders returns the finders used to load info files for all namespaces provided.
+	InfoFilesFinders() []InfoFilesFinder
+
+	// SetInstrumentOptions sets the instrument options.
+	SetInstrumentOptions(value instrument.Options) StateOptions
+
+	// InstrumentOptions returns the instrument options.
+	InstrumentOptions() instrument.Options
+}
+
+// InfoFilesFinder is used to lookup info files for the given combination of namespace and shards.
+type InfoFilesFinder struct {
```

Comment:
nit: Maybe we can name this something more generic? Or perhaps just use the `bootstrap.Namespace` struct here?
```
// Namespace is a namespace that is being bootstrapped.
type Namespace struct {
        // Metadata of the namespace being bootstrapped.
        Metadata namespace.Metadata
        // Shards is the shards for the namespace being bootstrapped.
        Shards []uint32
... other fields we don't care about
```
Reason being is that I'm planning on using this metadata for caching reading index info files as well.

Reply:
Ah, makes sense. I can take a crack at something more generic

Reply:
Ended up going w/ `NamespaceDetails`. Don't love it, but didn't love all the extra cruft on `bootstrap.Namespace` that could confuse someone into thinking it's useful for the `Cache`